### PR TITLE
feat: OpenTelemetry Tracing API and Exporting

### PR DESCRIPTION
**Pull request by devsnek**
Initial import of OTEL code supporting tracing. Metrics soon to come. Implements APIs for https://jsr.io/@deno/otel so that code using OpenTelemetry.js just works tm.

There is still a lot of work to do with configuration and adding built-in tracing to core APIs, which will come in followup PRs.

### CODE REVIEW

File: Cargo.lock [hunk 1]
```diff
@@ -1619,7 +1666,7 @@ dependencies = [
  "encoding_rs",
  "futures",
  "import_map",
- "indexmap",
+ "indexmap 2.3.0",
```
Comment (bartlomieju) [inline Cargo.lock:1669]: @devsnek could you try to remove so many duplicate versions of various dependencies added in this PR? Maybe we could bump some of the version we use in Deno?

File: cli/ops/bench.rs [hunk 2]
```diff
@@ -148,7 +148,7 @@ fn op_dispatch_bench_event(state: &mut OpState, #[serde] event: BenchEvent) {
 #[op2(fast)]
 #[number]
 fn op_bench_now(state: &mut OpState) -> Result<u64, AnyError> {
-  let ns = state.borrow::<time::Instant>().elapsed().as_nanos();
+  let ns = state.borrow::<StartTime>().elapsed().as_nanos();
```
Comment (bartlomieju) [inline cli/ops/bench.rs:151]: Is this a standalone fix that we could do in a separate PR maybe?

File: cli/tsc/dts/lib.deno.unstable.d.ts [hunk 3]
```diff
@@ -1225,6 +1225,103 @@ declare namespace Deno {
...
+       */
+      enter(): void;
+
+      /**
+       * Exit this span as the "current" span and restore the previous one.
+       */
+      exit(): void;
+
+      /**
+       * End this span, and exit it as the "current" span.
```
Comment (bartlomieju) [inline cli/tsc/dts/lib.deno.unstable.d.ts:1291]: What's the difference between `exit()` and `end()`. Does `end()` not restore the previous one?

File: cli/tsc/dts/lib.deno.unstable.d.ts [hunk 4]
```diff
@@ -1225,6 +1225,103 @@ declare namespace Deno {
...
+       */
+      end(): void;
+
+      [Symbol.dispose](): void;
+
+      static current(): Span | undefined | null;
```
Comment (bartlomieju) [inline cli/tsc/dts/lib.deno.unstable.d.ts:1297]: Is this on purpose that it's `undefined | null`?

File: cli/tsc/dts/lib.deno.unstable.d.ts [hunk 5]
```diff
@@ -1225,6 +1225,103 @@ declare namespace Deno {
...
+    }
+
+    /**
+     * A SpanExporter compatible with OpenTelemetry.js
```
Comment (bartlomieju) [inline cli/tsc/dts/lib.deno.unstable.d.ts:1301]: Could you link here and in the next one to relevant package or repo?

File: cli/tsc/dts/lib.deno.unstable.d.ts [hunk 6]
```diff
@@ -1225,6 +1225,103 @@ declare namespace Deno {
...
+    export {}; // only export exports
+  }
+
+  /**
+   * @category Telemetry
+   * @experimental
+   */
+  export namespace metrics {
+    export {}; // only export exports
+  }
```
Comment (bartlomieju) [inline cli/tsc/dts/lib.deno.unstable.d.ts:1323]: I don't understand this bit - what does `only export exports` mean here?

File: ext/http/00_serve.ts [hunk 7]
```diff
@@ -384,8 +388,10 @@ class CallbackContext {
   /** @type {Promise<void> | undefined} */
   closing;
   listener;
+  asyncContext;
 
   constructor(signal, args, listener) {
+    this.asyncContext = getAsyncContext();
```
Comment (bartlomieju) [inline ext/http/00_serve.ts:394]: Please remind me - does it mean that there's some overhead added no matter if OTEL is enabled or not?

File: ext/http/00_serve.ts [hunk 8]
```diff
@@ -495,82 +501,89 @@ function fastSyncResponseOrStream(
...
-      if (!ObjectPrototypeIsPrototypeOf(ResponsePrototype, response)) {
-        throw new TypeError(
-          "Return value from serve handler must be a response or a promise resolving to a response",
-        );
-      }
-
-      if (response.type === "error") {
-        throw new TypeError(
-          "Return value from serve handler must not be an error response (like Response.error())",
+    try {
```
Comment (bartlomieju) [inline ext/http/00_serve.ts:507]: Consider creating another function that is called inside `try {} finally {}` instead of inlining all of this logic here.

File: ext/web/15_performance.js [hunk 9]
```diff
@@ -90,8 +93,17 @@ webidl.converters["DOMString or PerformanceMeasureOptions"] = (
...
+const hr = new Uint32Array(TypedArrayPrototypeGetBuffer(hrU8));
+
+function setTimeOrigin() {
+  op_time_origin(hrU8);
+  timeOrigin = hr[0] * 1000 + hr[1] / 1e6;
+}
+
+function opNow() {
+  op_now(hrU8);
+  return (hr[0] * 1000 + hr[1] / 1e6);
```
Comment (bartlomieju) [inline ext/web/15_performance.js:106]: Ditto, consider doing this fix in a separate PR - will make it much easier to review and test in separation.

File: ext/web/benches/timers_ops.rs [hunk 10]
```diff
@@ -6,15 +6,6 @@ use deno_bench_util::bencher::benchmark_group;
 use deno_bench_util::bencher::Bencher;
 use deno_core::Extension;
 
-#[derive(Clone)]
-struct Permissions;
-
-impl deno_web::TimersPermission for Permissions {
```
Comment (bartlomieju) [inline ext/web/benches/timers_ops.rs:12]: Ugh, I think we forgot to remove these permissions when we deprecated `--allow-hrtime` flag :(

File: runtime/lib.rs [hunk 11]
```diff
@@ -99,38 +99,44 @@ pub static UNSTABLE_GRANULAR_FLAGS: &[UnstableGranularFlag] = &[
     show_in_help: true,
     id: 7,
   },
+  UnstableGranularFlag {
+    name: "otel",
+    help_text: "Enable unstable OpenTelemetry features",
+    show_in_help: false,
+    id: 8,
+  },
```
Comment (bartlomieju) [inline runtime/lib.rs:107]: Could have added this one at the end with id 12 to avoid all other changes 😛

File: runtime/ops/os/mod.rs [hunk 12]
```diff
@@ -186,6 +186,8 @@ fn op_get_exit_code(state: &mut OpState) -> i32 {
 
 #[op2(fast)]
 fn op_exit(state: &mut OpState) {
+  crate::ops::otel::otel_drop_state(state);
```
Comment (bartlomieju) [inline runtime/ops/os/mod.rs:189]: I assume this is gonna be blocking/sending some requests. Might be worth adding a comment here. What happens if the program exits "cleanly"? Is this a best effort solution or do we want to guarantee that all data is flushed/sent to the collector?

File: runtime/ops/otel.rs [hunk 13]
```diff
@@ -0,0 +1,582 @@
...
+      .into_iter()
+      .map(|(k, v)| KeyValue::new(k, v)),
+    ))
+  }
+
+  // The OTLP endpoint is automatically picked up from the
+  // `OTEL_EXPORTER_OTLP_ENDPOINT` environment variable. Additional headers can
+  // be specified using `OTEL_EXPORTER_OTLP_HEADERS`.
+
+  let client = reqwest::Client::new();
```
Comment (bartlomieju) [inline runtime/ops/otel.rs:263]: Could we try to use hyper directly? We made some efforts this year to move away from using `reqwest` in favor of `hyper` directly to have greater control. Maybe it's fine for the first pass, but if so I'd like to have an issue for a future TODO for that.

**Review by bartlomieju — COMMENTED** [review]
First pass from me. This looks really great 🚀

File: runtime/Cargo.toml [hunk 14]
```diff
@@ -114,8 +114,13 @@ log.workspace = true
 netif = "0.1.6"
 notify.workspace = true
 once_cell.workspace = true
+opentelemetry.workspace = true
+opentelemetry-otlp.workspace = true
+opentelemetry-semantic-conventions.workspace = true
+opentelemetry_sdk.workspace = true
 percent-encoding.workspace = true
 regex.workspace = true
+reqwest.workspace = true
```
Comment (piscisaureus) [inline runtime/Cargo.toml:123]: Ideally we wouldn't add back a dependency on reqwest, and use hyper directly.

File: runtime/js/telemetry.js [hunk 15]
```diff
@@ -0,0 +1,378 @@
...
+    }
+  }
+  return table;
+})();
+
+function generateId(bytes) {
+  let out = "";
+  for (let i = 0; i < bytes / 4; i += 1) {
+    const r32 = (MathRandom() * 2 ** 32) >>> 0;
+    out += hexSliceLookupTable[(r32 >> 24) & 0xff];
```
Comment (piscisaureus) [inline runtime/js/telemetry.js:52]: Was this somehow faster than `r32.toString(16).padStart(8, '0')` ?

File: runtime/ops/os/mod.rs (hunk 12 above)
Comment (piscisaureus) [inline runtime/ops/os/mod.rs:189]: We should have some tests that ensure the OTEL pipeline is properly flushed before the process exits. Some scenarios:

* `console.log("foo"); // Program naturally exits here`
* `console.log("foo"); Deno.exit(0);`
* `console.log("foo"); throw new Error("uncaught");`
* Web worker logs something  and immediate calls `self.close()`
* Web worker logs something and the main worker immediately calls `worker.terminate()`
* Web worker logs something and calls `Deno.exit()`
* Web worker logs something and throws an exception.

File: runtime/ops/otel.rs [hunk 16]
```diff
@@ -0,0 +1,582 @@
...
+}
+
+fn otel_create_globals(
+  config: OtelConfig,
+  op_state: &mut OpState,
+) -> anyhow::Result<()> {
+  // Parse the `OTEL_EXPORTER_OTLP_PROTOCOL` variable. The opentelemetry_*
+  // crates don't do this automatically.
+  // TODO(piscisaureus): enable GRPC support.
+  let protocol = match env::var("OTEL_EXPORTER_OTLP_PROTOCOL").as_deref() {
```
Comment (piscisaureus) [inline runtime/ops/otel.rs:214]: For a follow-up PR: we should have a way to specify OTLP protocol/endpoint/attributes via arguments rather than environment variables.

File: runtime/js/telemetry.js [hunk 17]
```diff
@@ -0,0 +1,378 @@
...
+      traceId = span.traceId;
+      spanId = span.spanId;
+      traceFlags = span.traceFlags;
+    }
+  }
+  return op_otel_log(message, level, traceId, spanId, traceFlags);
+}
+
+const otelConsoleConfig = {
+  ignore: 0,
```
Comment (piscisaureus) [inline runtime/js/telemetry.js:341]: I can see where this is going, but am I right that the only "mode" that is currently supported is "capture" ?

File: runtime/lib.rs (hunk 11 above)
Comment (devsnek) [reply runtime/lib.rs:107]: no they have to be in alphabetical order, there's a test for it.

File: runtime/ops/os/mod.rs (hunk 12 above)
Comment (devsnek) [reply runtime/ops/os/mod.rs:189]: honestly I'm not sure what's up with this function, i would expect everything in OpState to be dropped correctly anyway so I'm not sure why it's needed, i was planning to investigate it later.

File: cli/tsc/dts/lib.deno.unstable.d.ts (hunk 3 above)
Comment (devsnek) [reply cli/tsc/dts/lib.deno.unstable.d.ts:1291]: enter and exit control the span scope. end marks the end time of the span and submits it to be collected.

File: cli/tsc/dts/lib.deno.unstable.d.ts (hunk 6 above)
Comment (devsnek) [reply cli/tsc/dts/lib.deno.unstable.d.ts:1323]: all the namespaces here have this, idk what it does

File: ext/http/00_serve.ts (hunk 7 above)
Comment (devsnek) [reply ext/http/00_serve.ts:394]: technically yes but when we benchmarked it, it was not measurable.

File: runtime/lib.rs (hunk 11 above)
Comment (bartlomieju) [reply runtime/lib.rs:107]: oh interesting...

File: runtime/ops/otel.rs (hunk 16 above)
Comment (bartlomieju) [reply runtime/ops/otel.rs:214]: Is there a particular reason for that? I think having ability to configure using env vars and maybe in `deno.json` would be enough?

File: ext/web/benches/timers_ops.rs (hunk 10 above)
Comment (lucacasonato) [reply ext/web/benches/timers_ops.rs:12]: No this being here is intentional. Embedders must still have the ability to restrict timer granularity

Comment (bartlomieju) [reply ext/web/benches/timers_ops.rs:12]: I believe this is completely gone in this PR

File: runtime/ops/otel.rs (hunk 16 above)
Comment (lucacasonato) [reply runtime/ops/otel.rs:214]: I think only env vars is perfectly fine for now. Maybe something for a follow up if there is demand

File: ext/web/benches/timers_ops.rs (hunk 10 above)
Comment (lucacasonato) [reply ext/web/benches/timers_ops.rs:12]: Then please revert that @devsnek

File: cli/args/mod.rs [hunk 18]
```diff
@@ -1129,6 +1130,25 @@ impl CliOptions {
...
+      .unstable_config
+      .features
+      .contains(&String::from("otel"))
+    {
+      Some(OtelConfig {
+        default_service_name: Cow::Borrowed("deno"),
+        default_service_version: Cow::Borrowed(
+          crate::version::DENO_VERSION_INFO.deno,
+        ),
+        ..Default::default()
```
Comment (lucacasonato) [inline cli/args/mod.rs:1145]: This shouldn't be the service name. We should set these the keys `process.runtime.name` and `process.runtime.version`: https://opentelemetry.io/docs/specs/semconv/resource/process/#javascript-runtimes

File: runtime/ops/otel.rs [hunk 19]
```diff
@@ -0,0 +1,582 @@
...
+    }
+  };
+
+  // Define the resource attributes that will be attached to all log records.
+  // These attributes are sourced as follows (in order of precedence):
+  //   * The `service.name` attribute from the `OTEL_SERVICE_NAME` env var.
+  //   * Additional attributes from the `OTEL_RESOURCE_ATTRIBUTES` env var.
+  //   * Default attribute values defined here.
+  // TODO(piscisaureus): add more default attributes (e.g. script path).
+  let mut resource = Resource::default();
```
Comment (lucacasonato) [inline runtime/ops/otel.rs:240]: Can we use `opentelemetry_sdk::resource::SdkProvidedResourceDetector` to detect things from env vars?

File: runtime/ops/otel.rs (hunk 16 above)
Comment (lucacasonato) [reply runtime/ops/otel.rs:214]: This can be replaced with `ExportConfig::default().protocol`.

File: runtime/ops/otel.rs (hunk 13 above)
Comment (lucacasonato) [reply runtime/ops/otel.rs:263]: Should be a simple change: https://github.com/open-telemetry/opentelemetry-rust/blob/main/opentelemetry-http/src/lib.rs#L104

**Comment by KnorpelSenf** [comment]
Does this in any way go towards #22102?

File: runtime/ops/os/mod.rs (hunk 12 above)
Comment (piscisaureus) [reply runtime/ops/os/mod.rs:189]: Deno.exit() exits the process without running any destructor.

**Comment by lucacasonato** [comment]
@KnorpelSenf No, not really.

File: runtime/ops/otel.rs (hunk 16 above)
Comment (piscisaureus) [reply runtime/ops/otel.rs:214]: > Is there a particular reason for that? I think having ability to configure using env vars and maybe in `deno.json` would be enough?

The purpose would be to allow platforms (e.g. Deploy) to configure OTEL without exposing the internal configuration to the app.

Comment (piscisaureus) [reply runtime/ops/otel.rs:214]: > This can be replaced with `ExportConfig::default().protocol`.

It doesn't - opentelemetry-rust doesn't read the `OTEL_EXPORTER_OTLP_PROTOCOL` environment variable. It uses a fixed default based on enabled crate features.

File: runtime/ops/otel.rs (hunk 19 above)
Comment (piscisaureus) [reply runtime/ops/otel.rs:240]: `Resource::default` will automatically pick up on `OTEL_SERVICE_NAME` and `OTEL_RESOURCE_ATTRIBUTES`.

File: runtime/js/telemetry.js [hunk 20]
```diff
@@ -0,0 +1,378 @@
...
+    if (span.spanContext) {
+      const context = span.spanContext();
+      traceId = context.traceId;
+      spanId = context.spanId;
+      traceFlags = context.traceFlags;
+    } else {
+      traceId = span.traceId;
+      spanId = span.spanId;
+      traceFlags = span.traceFlags;
+    }
```
Comment (lucacasonato) [inline runtime/js/telemetry.js:335]: I don't get it - when is `spanContext` empty?

File: runtime/js/telemetry.js [hunk 21]
```diff
@@ -0,0 +1,378 @@
...
+  switch (consoleConfig) {
+    case otelConsoleConfig.capture:
+      core.wrapConsole(globalThis.console, new Console(otelLog));
+      break;
+    case otelConsoleConfig.replace:
+      ObjectDefineProperty(
+        globalThis,
+        "console",
+        core.propNonEnumerable(new Console(otelLog)),
+      );
```
Comment (lucacasonato) [inline runtime/js/telemetry.js:361]: It should probably still be wired into the original V8 console so that inspector console stays working.

File: cli/tsc/dts/lib.deno.unstable.d.ts [hunk 22]
```diff
@@ -1225,6 +1225,103 @@ declare namespace Deno {
...
+
+      /**
+       * Enter this span as the "current" span.
+       */
+      enter(): void;
+
+      /**
+       * Exit this span as the "current" span and restore the previous one.
+       */
+      exit(): void;
```
[the remark this answers is not in the document]
Comment (devsnek) [reply cli/tsc/dts/lib.deno.unstable.d.ts:1288]: otel-js has a pretty confusing api but at a high level it does not have the ability to enter/exit spans without calling a function, like this: `const span = tracer.startActiveSpan(() => { doSomething(); span.end(); });`. On our side we have enter/exit, but we also have the nicer choice of doing `using span = new Deno.tracing.Span();` which will automatically exit/end when disposed.

File: runtime/js/telemetry.js (hunk 21 above)
Comment (devsnek) [reply runtime/js/telemetry.js:361]: that happens in `99_main.js`.

File: runtime/js/telemetry.js (hunk 20 above)
Comment (devsnek) [reply runtime/js/telemetry.js:335]: originally when i wrote this our spans did not have `spanContext` for compat. i can reverse this to be `if (span.spanId)` to avoid calling spanContext unless needed.

Comment (lucacasonato) [reply runtime/js/telemetry.js:335]: yes that'd be good

File: cli/tsc/dts/lib.deno.unstable.d.ts (hunk 22 above)
[the remark this answers is not in the document]
Comment (lucacasonato) [reply cli/tsc/dts/lib.deno.unstable.d.ts:1288]: Ok - because we have that, do we need `enter` and `exit`? I think we should just force people to use `using`?

**Review by lucacasonato — APPROVED** [review]
LGTM once you resolve the two open threads

File: runtime/js/telemetry.js [hunk 23]
```diff
@@ -0,0 +1,378 @@
...
+
+    const parent = Span.current();
+    if (parent) {
+      if (parent.spanContext) {
+        const context = parent.spanContext();
+        this.parentSpanId = context.spanId;
+        this.traceId = context.traceId ?? generateId(TRACE_ID_BYTES);
+      } else {
+        this.parentSpanId = parent.spanId;
+        this.traceId = parent.traceId ?? generateId(TRACE_ID_BYTES);
```
Comment (lucacasonato) [inline runtime/js/telemetry.js:145]: why does this need to default to `generateId`?

File: runtime/ops/otel.rs (hunk 19 above)
Comment (tolu) [reply runtime/ops/otel.rs:331]: I'm currently trying to set up Jaeger for my deno app and it seems to me that neither of `OTEL_SERVICE_NAME ` or `OTEL_EXPORTER_OTLP_PROTOCOL` is picked up automatically 🤔 

Env file (`.env`)
```env
OTEL_DENO=true
OTEL_SERVICE_NAME=my-deno-app
OTEL_EXPORTER_OTLP_ENDPOINT=localhost:4318
OTEL_EXPORTER_OTLP_PROTOCOL=http/json
```

Starting deno (2.1.7) with:
```sh
deno run --env-file=./.env --unstable-otel -A main.ts
```

Printing `Deno.env` (with some key filtering) on program start yields:  
image


I then have a "logger" Deno app on `localhost:4318` that is printing:
```sh
 --> POST http://localhost:4318/v1/traces
application/x-protobuf

----------Read body-----------

�

�
!

service.name
unknown_service
%
telemetry.sdk.language

        deno-rust
"
process.runtime.version
2.1.7
```

So it seems like neither of the ENV variables are being picked up 🤔

Comment (devsnek) [reply runtime/ops/otel.rs:331]: @tolu could you please open a separate issue, thanks

Comment (tolu) [reply runtime/ops/otel.rs:331]: ofc 🙏

Comment (tolu) [reply runtime/ops/otel.rs:331]: https://github.com/denoland/deno/issues/27851
